File: android-app/app/src/main/java/lvt/crm/data/work/WorkRepository.kt
package lvt.crm.data.work

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import lvt.crm.data.convex.ConvexConfig
import lvt.crm.data.convex.ConvexException
import lvt.crm.data.convex.ConvexHttpClient
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.File
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

data class WorkUploadedEvidence(
    val driveFileId: String,
    val driveChecksum: String,
    val cleanupToken: String,
    val fileName: String,
    val fileType: String,
    val fileSize: Long,
)

data class WorkTaskItem(
    val id: String,
    val kind: Kind,
    val title: String,
    val deadline: String,
    val status: String,
    val documentContent: String,
    val departmentName: String,
    val qualityPercent: Int?,
    val rejectionReason: String,
    val isAdmin: Boolean,
    val documentTitle: String = "",
    val fileName: String = "",
    val memberNames: List<String> = emptyList(),
    val note: String = "",
) {
    enum class Kind { WorkItem, PersonalTask }
}

data class WorkMemberItem(
    val id: String,
    val name: String,
    val status: String,
)

data class WorkDocumentAssignment(
    val id: String,
    val departmentName: String,
    val content: String,
    val deadline: String,
    val status: String,
    val members: List<WorkMemberItem>,
)

data class WorkCompletionReviewItem(
    val workItemId: String,
    val userId: String,
    val userName: String,
    val content: String,
    val deadline: String,
    val departmentName: String,
    val note: String = "",
)

data class WorkApprovalItem(
    val id: String,
    val fileName: String,
    val content: String,
    val deadline: String,
    val status: String,
    val approvalCount: Int,
    val approvalTotal: Int,
    val myDecision: String,
    val assignments: List<WorkDocumentAssignment> = emptyList(),
    val fileUrl: String = "",
    val privateFile: Boolean = false,
)

data class WorkSnapshot(
    val assignerMode: String,
    val isAdmin: Boolean,
    val accessLevel: Int,
    val tasks: List<WorkTaskItem>,
    val approvals: List<WorkApprovalItem>,
    val completionReviews: List<WorkCompletionReviewItem> = emptyList(),
    val canCreate: Boolean = false,
    val isOps: Boolean = false,
)

internal data class ApprovalDecision(
    val userId: String,
    val approved: Boolean,
    val rejected: Boolean,
)

internal fun decisionForUser(currentUserId: String, approvers: List<ApprovalDecision>): String {
    val approver = approvers.firstOrNull { it.userId == currentUserId } ?: return ""
    return when {
        approver.approved -> "approved"
        approver.rejected -> "rejected"
        else -> ""
    }
}

interface WorkOperations {
    suspend fun listMine(): WorkSnapshot
    suspend fun complete(
        item: WorkTaskItem,
        qualityPercent: Int? = null,
        evidence: WorkUploadedEvidence? = null,
        note: String? = null,
    )
    suspend fun uploadEvidence(fileBytes: ByteArray, fileName: String, mimeType: String): WorkUploadedEvidence =
        throw UnsupportedOperationException()
    suspend fun decideApproval(documentId: String, approve: Boolean)
    suspend fun reviewCompletion(
        review: WorkCompletionReviewItem,
        approve: Boolean,
        qualityPercent: Int? = null,
        rejectionReason: String? = null,
    )
    suspend fun formOptions(): WorkFormOptions = throw UnsupportedOperationException()
    suspend fun createDocument(
        title: String,
        assignments: List<WorkCreateAssignment>,
        evidence: WorkUploadedEvidence? = null,
    ): String = throw UnsupportedOperationException()
}

class WorkRepository(
    private val convex: ConvexHttpClient,
    private val tokenProvider: () -> String? = { null },
    cacheDir: File? = null,
    private val webUrl: String = ConvexConfig.webUrl,
) : WorkOperations {
    private val documentCache = cacheDir?.let { WorkDocumentCache(File(it, "work-documents")) }
    private val downloadHttp = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(180, TimeUnit.SECONDS)
        .writeTimeout(45, TimeUnit.SECONDS)
        .followRedirects(true)
        .build()
    override suspend fun listMine(): WorkSnapshot {
        val result = convex.query("work:listMine")
        val isAdmin = result.optBoolean("isAdmin", false)
        val canCreate = result.optBoolean("canCreate", false)
        val accessLevel = result.optInt("level", 0)
        val assignerMode = result.optString("assignerMode", "")
        val currentUserId = result.optString("userId")
        val tasks = mutableListOf<WorkTaskItem>()
        val approvals = mutableListOf<WorkApprovalItem>()
        var isOps = isAdmin

        val approvalArray = result.optJSONArray("approvals")
        if (approvalArray != null) {
            for (i in 0 until approvalArray.length()) {
                val document = approvalArray.getJSONObject(i)
                val approvers = document.optJSONArray("approvers") ?: org.json.JSONArray()
                val decisions = mutableListOf<ApprovalDecision>()
                for (approverIndex in 0 until approvers.length()) {
                    val approver = approvers.optJSONObject(approverIndex) ?: continue
                    decisions += ApprovalDecision(
                        userId = approver.optString("_id"),
                        approved = approver.optBoolean("approved", false),
                        rejected = approver.optBoolean("rejected", false),
                    )
                }
                val myDecision = decisionForUser(currentUserId, decisions)
                approvals += WorkApprovalItem(
                    id = document.optString("_id"),
                    fileName = document.optString("fileName"),
                    content = document.optString("content"),
                    deadline = document.optString("deadline"),
                    status = document.optString("status"),
                    approvalCount = document.optInt("approvalCount"),
                    approvalTotal = document.optInt("approvalTotal"),
                    myDecision = myDecision,
                    assignments = parseAssignments(document.optJSONArray("assignments")),
                    fileUrl = document.optString("fileUrl"),
                    privateFile = document.optBoolean("privateFile", false),
                )
            }
        }

        var completionReviews = emptyList<WorkCompletionReviewItem>()
        if (isAdmin || canCreate) {
            val adminResult = convex.query("work:listAdmin")
            isOps = adminResult.optBoolean("isOps", isAdmin)
            approvals.clear()
            val documents = adminResult.optJSONArray("documents") ?: org.json.JSONArray()
            for (i in 0 until documents.length()) {
                val document = documents.getJSONObject(i)
                approvals += WorkApprovalItem(
                    id = document.optString("_id"),
                    fileName = document.optString("fileName"),
                    content = document.optString("content"),
                    deadline = document.optString("deadline"),
                    status = document.optString("status"),
                    approvalCount = document.optInt("approvalCount"),
                    approvalTotal = document.optInt("approvalTotal"),
                    myDecision = "",
                    assignments = parseAssignments(document.optJSONArray("assignments")),
                    fileUrl = document.optString("fileUrl"),
                    privateFile = document.optBoolean("privateFile", false),
                )
            }
            completionReviews = parseCompletionReviews(adminResult.optJSONArray("pendingCompletionReviews"))
        } else {
            completionReviews = parseCompletionReviews(result.optJSONArray("pendingCompletionReviews"))
        }

        val myTasks = result.optJSONArray("myTasks")
        if (myTasks != null) {
            for (i in 0 until myTasks.length()) {
                val t = myTasks.getJSONObject(i)
                tasks += WorkTaskItem(
                    id = t.optString("_id"),
                    kind = WorkTaskItem.Kind.WorkItem,
                    title = t.optString("content"),
                    deadline = t.optString("deadline"),
                    status = t.optString("status"),
                    documentContent = t.optString("documentContent"),
                    departmentName = t.optString("departmentName"),
                    qualityPercent = t.optIntOrNull("qualityPercent"),
                    rejectionReason = t.optString("rejectionReason"),
                    isAdmin = isAdmin,
                    documentTitle = t.optString("documentTitle"),
                    fileName = t.optString("fileName"),
                    memberNames = t.optJSONArray("members").toMemberNames(),
                    note = t.optWorkNote(),
                )
            }
        }

        val personal = result.optJSONArray("personalTasks")
        if (personal != null) {
            for (i in 0 until personal.length()) {
                val t = personal.getJSONObject(i)
                tasks += WorkTaskItem(
                    id = t.optString("_id"),
                    kind = WorkTaskItem.Kind.PersonalTask,
                    title = t.optString("title").ifBlank { t.optString("documentContent") },
                    deadline = t.optString("deadline"),
                    status = t.optString("status"),
                    documentContent = t.optString("documentContent"),
                    departmentName = t.optString("departmentName"),
                    qualityPercent = t.optIntOrNull("qualityPercent"),
                    rejectionReason = t.optString("rejectionReason"),
                    isAdmin = isAdmin,
                    documentTitle = t.optString("documentTitle"),
                    fileName = t.optString("fileName"),
                    memberNames = t.optJSONArray("assignees").toMemberNames(),
                    note = t.optWorkNote(),
                )
            }
        }

        return WorkSnapshot(
            assignerMode = assignerMode,
            isAdmin = isAdmin,
            accessLevel = accessLevel,
            tasks = orderedWorkTasks(tasks),
            approvals = approvals.sortedWith(compareBy<WorkApprovalItem>({ it.status != "pending" }, { it.deadline })),
            completionReviews = completionReviews,
            canCreate = canCreate,
            isOps = isOps,
        )
    }

    override suspend fun formOptions(): WorkFormOptions {
        val result = convex.query("work:formOptions")
        val departments = result.optJSONArray("departments") ?: org.json.JSONArray()
        val users = result.optJSONArray("users") ?: org.json.JSONArray()
        return WorkFormOptions(
            canCreate = result.optBoolean("canCreate", false),
            isOps = result.optBoolean("isOps", false),
            departments = List(departments.length()) { index ->
                val item = departments.getJSONObject(index)
                WorkFormDepartment(
                    id = item.optString("_id"),
                    name = item.optString("name"),
                )
            },
            users = List(users.length()) { index ->
                val item = users.getJSONObject(index)
                WorkFormUser(
                    id = item.optString("_id"),
                    name = item.optString("name").ifBlank { item.optString("email") },
                    departmentName = item.optString("departmentName"),
                    level = item.optInt("level", 0),
                )
            },
        )
    }

    override suspend fun createDocument(
        title: String,
        assignments: List<WorkCreateAssignment>,
        evidence: WorkUploadedEvidence?,
    ): String {
        val args = org.json.JSONObject()
            .put("title", title.trim())
            .put("assignments", org.json.JSONArray().apply {
                assignments.forEach { row ->
                    put(
                        org.json.JSONObject().apply {
                            put("type", row.type)
                            put("content", row.content.trim())
                            put("deadline", row.deadline.trim())
                            if (row.isIndividual) {
                                put("userIds", org.json.JSONArray(row.userIds.filter { it.isNotBlank() }))
                            } else {
                                put("departmentId", row.departmentId)
                            }
                        },
                    )
                }
            })
        if (evidence != null) {
            args.put("driveFileId", evidence.driveFileId)
            args.put("driveChecksum", evidence.driveChecksum)
            args.put("cleanupToken", evidence.cleanupToken)
            args.put("fileName", evidence.fileName)
            args.put("fileType", evidence.fileType)
            args.put("fileSize", evidence.fileSize)
        }
        return try {
            val result = convex.mutation("work:createDocument", args)
            if (evidence != null && evidence.cleanupToken.isNotBlank()) {
                settleUploadedFile(evidence.cleanupToken, finalize = true)
            }
            result.optString("value").ifBlank { result.optString("_id") }
        } catch (error: Exception) {
            if (evidence != null && evidence.cleanupToken.isNotBlank()) {
                settleUploadedFile(evidence.cleanupToken, finalize = false)
            }
            throw error
        }
    }

    override suspend fun complete(item: WorkTaskItem, qualityPercent: Int?, evidence: WorkUploadedEvidence?, note: String?) {
        val args = JSONObject()
        if (qualityPercent != null) {
            args.put("qualityPercent", qualityPercent)
        }
        if (evidence != null) {
            args.put("driveFileId", evidence.driveFileId)
            args.put("driveChecksum", evidence.driveChecksum)
            args.put("cleanupToken", evidence.cleanupToken)
            args.put("fileName", evidence.fileName)
            args.put("fileType", evidence.fileType)
            args.put("fileSize", evidence.fileSize)
        }
        val trimmedNote = note?.trim().orEmpty()
        if (trimmedNote.isNotEmpty()) {
            args.put("note", trimmedNote.take(500))
        }
        when (item.kind) {
            WorkTaskItem.Kind.WorkItem -> {
                args.put("workItemId", item.id)
                convex.mutation("work:completeWorkItem", args)
            }
            WorkTaskItem.Kind.PersonalTask -> {
                args.put("taskId", item.id)
                convex.mutation("work:completePersonalTask", args)
            }
        }
        if (evidence != null && evidence.cleanupToken.isNotBlank()) {
            settleUploadedFile(evidence.cleanupToken, finalize = true)
        }
    }

    override suspend fun uploadEvidence(fileBytes: ByteArray, fileName: String, mimeType: String): WorkUploadedEvidence = withContext(Dispatchers.IO) {
        if (fileBytes.isEmpty()) {
            throw ConvexException("INVALID_FILE", "Tệp đính kèm không có nội dung.")
        }
        if (fileBytes.size > 20 * 1024 * 1024) {
            throw ConvexException("FILE_TOO_LARGE", "Dung lượng tệp vượt quá giới hạn 20MB.")
        }
        val token = tokenProvider()?.takeIf { it.isNotBlank() }
            ?: throw ConvexException("AUTH_REQUIRED", "Vui lòng đăng nhập lại.")
        val base = webUrl.trimEnd('/')
        val encodedName = URLEncoder.encode(fileName, Charsets.UTF_8.name()).replace("+", "%20")
        val mediaType = mimeType.toMediaTypeOrNull() ?: "application/octet-stream".toMediaType()
        val request = Request.Builder()
            .url("$base/api/files/upload")
            .header("Authorization", "Bearer $token")
            .header("Content-Type", mimeType)
            .header("X-File-Name", encodedName)
            .header("X-LVT-Upload-Purpose", "work")
            .post(fileBytes.toRequestBody(mediaType))
            .build()

        downloadHttp.newCall(request).execute().use { response ->
            val body = response.body?.string().orEmpty()
            if (!response.isSuccessful) {
                val errorJson = runCatching { JSONObject(body) }.getOrNull()
                val code = errorJson?.optString("error")?.takeIf { it.isNotBlank() } ?: "WORK_UPLOAD_FAILED:${response.code}"
                throw ConvexException(code, "Không thể tải bằng chứng lên. Vui lòng thử lại.")
            }
            val json = runCatching { JSONObject(body) }.getOrElse {
                throw ConvexException("WORK_UPLOAD_INVALID_RESPONSE", "Phản hồi tải tệp không hợp lệ.")
            }
            val driveFileId = json.optString("driveFileId")
            val cleanupToken = json.optString("cleanupToken")
            if (driveFileId.isBlank() || cleanupToken.isBlank()) {
                throw ConvexException("WORK_UPLOAD_INVALID_RESPONSE", "Phản hồi tải tệp không hợp lệ.")
            }
            val driveChecksum = json.optString("driveChecksum")
            WorkUploadedEvidence(
                driveFileId = driveFileId,
                driveChecksum = driveChecksum,
                cleanupToken = cleanupToken,
                fileName = fileName,
                fileType = mimeType,
                fileSize = fileBytes.size.toLong(),
            )
        }
    }

    private suspend fun settleUploadedFile(cleanupToken: String, finalize: Boolean) = withContext(Dispatchers.IO) {
        val token = tokenProvider()?.takeIf { it.isNotBlank() } ?: return@withContext
        val encodedToken = URLEncoder.encode(cleanupToken, Charsets.UTF_8.name()).replace("+", "%20")
        val base = webUrl.trimEnd('/')
        val request = Request.Builder()
            .url("$base/api/files/uploads/$encodedToken")
            .header("Authorization", "Bearer $token")
            .method(if (finalize) "POST" else "DELETE", "".toRequestBody("text/plain".toMediaType()))
            .build()
        runCatching { downloadHttp.newCall(request).execute().close() }
    }

    override suspend fun decideApproval(documentId: String, approve: Boolean) {
        val path = if (approve) "work:approveDocument" else "work:rejectDocument"
        convex.mutation(path, JSONObject().put("documentId", documentId))
    }

    override suspend fun reviewCompletion(
        review: WorkCompletionReviewItem,
        approve: Boolean,
        qualityPercent: Int?,
        rejectionReason: String?,
    ) {
        val args = JSONObject()
            .put("workItemId", review.workItemId)
            .put("userId", review.userId)
            .put("decision", if (approve) "approve" else "reject")
        if (qualityPercent != null) args.put("qualityPercent", qualityPercent)
        if (!rejectionReason.isNullOrBlank()) args.put("rejectionReason", rejectionReason)
        convex.mutation("work:reviewWorkCompletion", args)
    }

    suspend fun downloadDocument(document: WorkApprovalItem): File = withContext(Dispatchers.IO) {
        val cache = documentCache
            ?: throw ConvexException("WORK_FILE_UNAVAILABLE", "Tệp công văn chưa sẵn sàng để mở.")
        val request: Request
        val sourceIdentity: String
        val publicUrl = document.fileUrl.trim()
        if (publicUrl.isNotEmpty()) {
            request = Request.Builder().url(publicUrl).get().build()
            sourceIdentity = "public:${document.id}:$publicUrl"
        } else if (document.privateFile) {
            val token = tokenProvider()?.takeIf { it.isNotBlank() }
                ?: throw ConvexException("WORK_FILE_FORBIDDEN", "Bạn không còn quyền mở tệp công văn này.")
            val encodedId = URLEncoder.encode(document.id, Charsets.UTF_8.name()).replace("+", "%20")
            val base = webUrl.trimEnd('/')
            val metadataRequest = Request.Builder()
                .url("$base/api/files/$encodedId/metadata")
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            val metadata = downloadHttp.newCall(metadataRequest).execute().use { response ->
                val body = response.body?.string().orEmpty()
                if (!response.isSuccessful) {
                    throw ConvexException("WORK_FILE_FORBIDDEN", "Bạn không còn quyền mở tệp công văn này.")
                }
                runCatching { JSONObject(body) }.getOrElse {
                    throw ConvexException("WORK_FILE_FORBIDDEN", "Bạn không còn quyền mở tệp công văn này.")
                }
            }
            val fileVersion = metadata.optString("fileVersion")
            if (fileVersion.isBlank()) {
                throw ConvexException("WORK_FILE_FORBIDDEN", "Bạn không còn quyền mở tệp công văn này.")
            }
            request = Request.Builder()
                .url("$base/api/files/$encodedId")
                .header("Authorization", "Bearer $token")
                .get()
                .build()
            sourceIdentity = "private:${document.id}:$fileVersion"
        } else {
            throw ConvexException("WORK_FILE_UNAVAILABLE", "Tệp công văn chưa sẵn sàng để mở.")
        }
        cache.cachedFile(sourceIdentity)?.let { cached ->
            return@withContext cached
        }
        val tempDir = cacheDirFallback().apply { mkdirs() }
        val downloaded = File.createTempFile("lvt-work-", ".part", tempDir)
        try {
            downloadHttp.newCall(request).execute().use { response ->
                if (!response.isSuccessful) {
                    throw ConvexException("WORK_FILE_DOWNLOAD_FAILED", "Không thể tải tệp công văn. Hãy thử lại.")
                }
                val body = response.body
                    ?: throw ConvexException("WORK_FILE_DOWNLOAD_FAILED", "Không thể tải tệp công văn. Hãy thử lại.")
                body.byteStream().use { input ->
                    downloaded.outputStream().use { output -> input.copyTo(output) }
                }
            }
            if (downloaded.length() <= 0L) {
                throw ConvexException("WORK_FILE_DOWNLOAD_FAILED", "Không thể tải tệp công văn. Hãy thử lại.")
            }
            return@withContext cache.store(
                downloaded,
                sourceIdentity,
                document.fileName.ifBlank { "cong-van" },
            )
        } catch (error: ConvexException) {
            downloaded.delete()
            throw error
        } catch (_: Exception) {
            downloaded.delete()
            throw ConvexException("WORK_FILE_DOWNLOAD_FAILED", "Không thể tải tệp công văn. Hãy thử lại.")
        }
    }

    private fun cacheDirFallback(): File = File(System.getProperty("java.io.tmpdir") ?: ".", "lvt-work-temp")
}

private fun org.json.JSONArray?.toMemberNames(): List<String> {
    if (this == null) return emptyList()
    return List(length()) { index ->
        val member = optJSONObject(index) ?: return@List ""
        member.optString("name").ifBlank { member.optString("email") }
    }.filter { it.isNotBlank() }
}

private fun parseAssignments(items: org.json.JSONArray?): List<WorkDocumentAssignment> {
    if (items == null) return emptyList()
    return List(items.length()) { index ->
        val item = items.getJSONObject(index)
        val members = item.optJSONArray("members") ?: org.json.JSONArray()
        WorkDocumentAssignment(
            id = item.optString("_id"),
            departmentName = item.optString("departmentName"),
            content = item.optString("content"),
            deadline = item.optString("deadline"),
            status = item.optString("status"),
            members = List(members.length()) { memberIndex ->
                val member = members.getJSONObject(memberIndex)
                WorkMemberItem(
                    id = member.optString("_id"),
                    name = member.optString("name").ifBlank { member.optString("email") },
                    status = member.optString("status"),
                )
            },
        )
    }
}

private fun parseCompletionReviews(items: org.json.JSONArray?): List<WorkCompletionReviewItem> {
    if (items == null) return emptyList()
    return List(items.length()) { index ->
        val item = items.getJSONObject(index)
        WorkCompletionReviewItem(
            workItemId = item.optString("workItemId"),
            userId = item.optString("userId"),
            userName = item.optString("userName"),
            content = item.optString("content"),
            deadline = item.optString("deadline"),
            departmentName = item.optString("departmentName"),
            note = item.optString("note").trim(),
        )
    }
}

internal fun needsCompletion(status: String): Boolean = status in setOf(
    "pending_task",
    "pending",
    "overdue",
    "rejected",
    "rejected_completion",
)

internal fun orderedWorkTasks(tasks: List<WorkTaskItem>): List<WorkTaskItem> = tasks.sortedWith(
    compareBy<WorkTaskItem>(
        { !needsCompletion(it.status) },
        { it.deadline },
        { it.title },
    ),
)

private fun JSONObject.optWorkNote(): String {
    val nested = optJSONObject("completion")?.optString("note")?.trim().orEmpty()
    if (nested.isNotEmpty()) return nested
    return optString("note").trim()
}

private fun JSONObject.optIntOrNull(key: String): Int? {
    if (!has(key) || isNull(key)) return null
    return optInt(key)
}

File: ios-app/LvtCrm/Data/Work/WorkRepository.swift
import Foundation

struct WorkTaskItem: Identifiable, Equatable, Sendable {
    enum Kind: String, Sendable { case workItem, personalTask }
    let id: String
    let kind: Kind
    let title: String
    let deadline: String
    var status: String
    let documentContent: String
    let departmentName: String
    let qualityPercent: Int?
    let rejectionReason: String
    let isAdmin: Bool
}

struct WorkMemberItem: Identifiable, Equatable, Sendable {
    let id: String
    let name: String
    let status: String
}

struct WorkDocumentAssignment: Identifiable, Equatable, Sendable {
    let id: String
    let departmentName: String
    let content: String
    let deadline: String
    let status: String
    let members: [WorkMemberItem]
}

struct WorkCompletionReviewItem: Identifiable, Equatable, Sendable {
    var id: String { workItemId + userId }
    let workItemId: String
    let userId: String
    let userName: String
    let content: String
    let deadline: String
    let departmentName: String
}

struct WorkApprovalItem: Identifiable, Equatable, Sendable {
    let id: String
    let fileName: String
    let content: String
    let deadline: String
    let status: String
    let approvalCount: Int
    let approvalTotal: Int
    var myDecision: String
    let assignments: [WorkDocumentAssignment]
}

struct WorkSnapshot: Equatable, Sendable {
    let assignerMode: String
    let isAdmin: Bool
    let accessLevel: Int
    let tasks: [WorkTaskItem]
    let approvals: [WorkApprovalItem]
    let completionReviews: [WorkCompletionReviewItem]
}

enum WorkHelpers {
    static func needsCompletion(_ status: String) -> Bool {
        ["pending_task", "pending", "overdue", "rejected", "rejected_completion"].contains(status)
    }

    static func orderedTasks(_ tasks: [WorkTaskItem]) -> [WorkTaskItem] {
        tasks.sorted {
            let lhsNeeds = needsCompletion($0.status)
            let rhsNeeds = needsCompletion($1.status)
            if lhsNeeds != rhsNeeds { return lhsNeeds && !rhsNeeds }
            if $0.deadline != $1.deadline { return $0.deadline < $1.deadline }
            return $0.title < $1.title
        }
    }

    static func decisionForUser(currentUserId: String, approvers: [[String: Any]]) -> String {
        guard let approver = approvers.first(where: { ($0["_id"] as? String) == currentUserId }) else {
            return ""
        }
        if (approver["approved"] as? Bool) == true { return "approved" }
        if (approver["rejected"] as? Bool) == true { return "rejected" }
        return ""
    }
}

final class WorkRepository: Sendable {
    private let convex: ConvexHttpClient

    init(convex: ConvexHttpClient) {
        self.convex = convex
    }

    func listMine() async throws -> WorkSnapshot {
        let result = try await convex.query("work:listMine")
        let isAdmin = (result["isAdmin"] as? Bool) ?? false
        let accessLevel = (result["level"] as? Int) ?? 0
        let assignerMode = (result["assignerMode"] as? String) ?? ""
        let currentUserId = (result["userId"] as? String) ?? ""

        var approvals: [WorkApprovalItem] = []
        var completionReviews: [WorkCompletionReviewItem] = []

        if isAdmin {
            let adminResult = try await convex.query("work:listAdmin")
            let documents = adminResult["documents"] as? [[String: Any]] ?? []
            approvals = documents.map { document in
                WorkApprovalItem(
                    id: (document["_id"] as? String) ?? "",
                    fileName: (document["fileName"] as? String) ?? "",
                    content: (document["content"] as? String) ?? "",
                    deadline: (document["deadline"] as? String) ?? "",
                    status: (document["status"] as? String) ?? "",
                    approvalCount: (document["approvalCount"] as? Int) ?? 0,
                    approvalTotal: (document["approvalTotal"] as? Int) ?? 0,
                    myDecision: "",
                    assignments: parseAssignments(document["assignments"] as? [[String: Any]])
                )
            }
            completionReviews = parseCompletionReviews(
                adminResult["pendingCompletionReviews"] as? [[String: Any]]
            )
        } else {
            let approvalArray = result["approvals"] as? [[String: Any]] ?? []
            approvals = approvalArray.map { document in
                let approvers = document["approvers"] as? [[String: Any]] ?? []
                return WorkApprovalItem(
                    id: (document["_id"] as? String) ?? "",
                    fileName: (document["fileName"] as? String) ?? "",
                    content: (document["content"] as? String) ?? "",
                    deadline: (document["deadline"] as? String) ?? "",
                    status: (document["status"] as? String) ?? "",
                    approvalCount: (document["approvalCount"] as? Int) ?? 0,
                    approvalTotal: (document["approvalTotal"] as? Int) ?? 0,
                    myDecision: WorkHelpers.decisionForUser(
                        currentUserId: currentUserId,
                        approvers: approvers
                    ),
                    assignments: parseAssignments(document["assignments"] as? [[String: Any]])
                )
            }
        }

        var tasks: [WorkTaskItem] = []
        for t in result["myTasks"] as? [[String: Any]] ?? [] {
            tasks.append(
                WorkTaskItem(
                    id: (t["_id"] as? String) ?? "",
                    kind: .workItem,
                    title: (t["content"] as? String) ?? "",
                    deadline: (t["deadline"] as? String) ?? "",
                    status: (t["status"] as? String) ?? "",
                    documentContent: (t["documentContent"] as? String) ?? "",
                    departmentName: (t["departmentName"] as? String) ?? "",
                    qualityPercent: t["qualityPercent"] as? Int,
                    rejectionReason: (t["rejectionReason"] as? String) ?? "",
                    isAdmin: isAdmin
                )
            )
        }
        for t in result["personalTasks"] as? [[String: Any]] ?? [] {
            let title = ((t["title"] as? String)?.nilIfBlank)
                ?? (t["documentContent"] as? String)
                ?? ""
            tasks.append(
                WorkTaskItem(
                    id: (t["_id"] as? String) ?? "",
                    kind: .personalTask,
                    title: title,
                    deadline: (t["deadline"] as? String) ?? "",
                    status: (t["status"] as? String) ?? "",
                    documentContent: (t["documentContent"] as? String) ?? "",
                    departmentName: (t["departmentName"] as? String) ?? "",
                    qualityPercent: t["qualityPercent"] as? Int,
                    rejectionReason: (t["rejectionReason"] as? String) ?? "",
                    isAdmin: isAdmin
                )
            )
        }

        let sortedApprovals = approvals.sorted {
            let lhsPending = $0.status == "pending"
            let rhsPending = $1.status == "pending"
            if lhsPending != rhsPending { return lhsPending && !rhsPending }
            return $0.deadline < $1.deadline
        }

        return WorkSnapshot(
            assignerMode: assignerMode,
            isAdmin: isAdmin,
            accessLevel: accessLevel,
            tasks: WorkHelpers.orderedTasks(tasks),
            approvals: sortedApprovals,
            completionReviews: completionReviews
        )
    }

    func complete(item: WorkTaskItem, qualityPercent: Int? = nil) async throws {
        var args: [String: Any] = [:]
        if let qualityPercent { args["qualityPercent"] = qualityPercent }
        switch item.kind {
        case .workItem:
            args["workItemId"] = item.id
            _ = try await convex.mutation("work:completeWorkItem", args: args)
        case .personalTask:
            args["taskId"] = item.id
            _ = try await convex.mutation("work:completePersonalTask", args: args)
        }
    }

    func decideApproval(documentId: String, approve: Bool) async throws {
        let path = approve ? "work:approveDocument" : "work:rejectDocument"
        _ = try await convex.mutation(path, args: ["documentId": documentId])
    }

    func reviewCompletion(
        review: WorkCompletionReviewItem,
        approve: Bool,
        qualityPercent: Int? = nil,
        rejectionReason: String? = nil
    ) async throws {
        var args: [String: Any] = [
            "workItemId": review.workItemId,
            "userId": review.userId,
            "decision": approve ? "approve" : "reject",
        ]
        if let qualityPercent { args["qualityPercent"] = qualityPercent }
        if let rejectionReason, !rejectionReason.isEmpty {
            args["rejectionReason"] = rejectionReason
        }
        _ = try await convex.mutation("work:reviewWorkCompletion", args: args)
    }

    private func parseAssignments(_ items: [[String: Any]]?) -> [WorkDocumentAssignment] {
        (items ?? []).map { item in
            let members = (item["members"] as? [[String: Any]] ?? []).map { member in
                WorkMemberItem(
                    id: (member["_id"] as? String) ?? "",
                    name: ((member["name"] as? String)?.nilIfBlank)
                        ?? (member["email"] as? String)
                        ?? "",
                    status: (member["status"] as? String) ?? ""
                )
            }
            return WorkDocumentAssignment(
                id: (item["_id"] as? String) ?? "",
                departmentName: (item["departmentName"] as? String) ?? "",
                content: (item["content"] as? String) ?? "",
                deadline: (item["deadline"] as? String) ?? "",
                status: (item["status"] as? String) ?? "",
                members: members
            )
        }
    }

    private func parseCompletionReviews(_ items: [[String: Any]]?) -> [WorkCompletionReviewItem] {
        (items ?? []).map { item in
            WorkCompletionReviewItem(
                workItemId: (item["workItemId"] as? String) ?? "",
                userId: (item["userId"] as? String) ?? "",
                userName: (item["userName"] as? String) ?? "",
                content: (item["content"] as? String) ?? "",
                deadline: (item["deadline"] as? String) ?? "",
                departmentName: (item["departmentName"] as? String) ?? ""
            )
        }
    }
}

private extension String {
    var nilIfBlank: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}
